CMakeLists.txt

```diff
@@ -892,125 +891,137 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
-  src/waveform/widgets/emptywaveformwidget.cpp
-  src/waveform/widgets/glrgbwaveformwidget.cpp
-  src/waveform/widgets/glsimplewaveformwidget.cpp
-  src/waveform/widgets/glslwaveformwidget.cpp
-  src/waveform/widgets/glvsynctestwidget.cpp
-  src/waveform/widgets/glwaveformwidget.cpp
-  src/waveform/widgets/hsvwaveformwidget.cpp
-  src/waveform/widgets/qthsvwaveformwidget.cpp
-  src/waveform/widgets/qtrgbwaveformwidget.cpp
-  src/waveform/widgets/qtsimplewaveformwidget.cpp
-  src/waveform/widgets/qtvsynctestwidget.cpp
-  src/waveform/widgets/qtwaveformwidget.cpp
-  src/waveform/widgets/rgbwaveformwidget.cpp
-  src/waveform/widgets/softwarewaveformwidget.cpp
-  src/waveform/widgets/waveformwidgetabstract.cpp
-  src/widget/controlwidgetconnection.cpp
-  src/widget/hexspinbox.cpp
-  src/widget/paintable.cpp
-  src/widget/wanalysislibrarytableview.cpp
-  src/widget/wbasewidget.cpp
-  src/widget/wbattery.cpp
-  src/widget/wbeatspinbox.cpp
-  src/widget/wcolorpicker.cpp
-  src/widget/wcolorpickeraction.cpp
-  src/widget/wcombobox.cpp
-  src/widget/wcoverart.cpp
-  src/widget/wcoverartlabel.cpp
-  src/widget/wcoverartmenu.cpp
-  src/widget/wcuemenupopup.cpp
-  src/widget/wdisplay.cpp
-  src/widget/weffect.cpp
-  src/widget/weffectbuttonparameter.cpp
-  src/widget/weffectchain.cpp
-  src/widget/weffectparameter.cpp
-  src/widget/weffectparameterbase.cpp
-  src/widget/weffectparameterknob.cpp
-  src/widget/weffectparameterknobcomposed.cpp
-  src/widget/weffectpushbutton.cpp
-  src/widget/weffectselector.cpp
-  src/widget/whotcuebutton.cpp
-  src/widget/wimagestore.cpp
-  src/widget/wkey.cpp
-  src/widget/wknob.cpp
-  src/widget/wknobcomposed.cpp
-  src/widget/wlabel.cpp
-  src/widget/wlibrary.cpp
-  src/widget/wlibrarysidebar.cpp
-  src/widget/wlibrarytableview.cpp
-  src/widget/wlibrarytextbrowser.cpp
-  src/widget/wmainmenubar.cpp
-  src/widget/wnumber.cpp
-  src/widget/wnumberdb.cpp
-  src/widget/wnumberpos.cpp
-  src/widget/wnumberrate.cpp
-  src/widget/woverview.cpp
-  src/widget/woverviewhsv.cpp
-  src/widget/woverviewlmh.cpp
-  src/widget/woverviewrgb.cpp
-  src/widget/wpixmapstore.cpp
-  src/widget/wpushbutton.cpp
-  src/widget/wrecordingduration.cpp
-  src/widget/wscrollable.cpp
-  src/widget/wsearchlineedit.cpp
-  src/widget/wsearchrelatedtracksmenu.cpp
-  src/widget/wsingletoncontainer.cpp
-  src/widget/wsizeawarestack.cpp
-  src/widget/wskincolor.cpp
-  src/widget/wslidercomposed.cpp
-  src/widget/wspinny.cpp
-  src/widget/wsplitter.cpp
-  src/widget/wstarrating.cpp
-  src/widget/wstatuslight.cpp
-  src/widget/wtime.cpp
-  src/widget/wtrackmenu.cpp
-  src/widget/wtrackproperty.cpp
-  src/widget/wtracktableview.cpp
-  src/widget/wtracktableviewheader.cpp
-  src/widget/wtracktext.cpp
-  src/widget/wtrackwidgetgroup.cpp
-  src/widget/wvumeter.cpp
-  src/widget/wwaveformviewer.cpp
-  src/widget/wwidget.cpp
-  src/widget/wwidgetgroup.cpp
-  src/widget/wwidgetstack.cpp
-  src/widget/wraterange.cpp
 )
+if(NOT QT6)
```

Comment:
Do we really need to remove all these files? Can't we keep everything except QGLWidget?

Reply:
I would also suggest to keep everything that builds fine with Qt5 and only remove files conditionally if incompatibilities arise or if they become obsolete.

Reply:
No, there is no point trying to get the legacy skin system to build with Qt6 because essential features cannot work with Qt6. That would probably require lots more preprocessor Qt version checks scattered around the legacy code.

Reply:
Which ones won't work? I though only QOpenGLWidget?

Reply:
The legacy skin system uses QGLWidget, not QOpenGLWidget, which is the reason it needs to be replaced.

Reply:
Yes, and if that is the only thing not working, we should just disable waveforms and spinnies I guess.

Reply:
Why?

Reply:
I think you're underestimating how much work there is left to do for QML support. Breaking legacy skins on Qt6 for no reason is a good idea if we can as well disable QGLWidget to make it build.

If at some point we need to do a release with Qt6 support and QML is not ready, we still have the option to implement QOpenGLWidget support instead.
Also, we still have the option to keep legacy skins for a transitional period.

Reply:
This is pointless. We're not going to release a GUI without waveforms and we're not going to use QOpenGLWidget.

Reply:
I agree with @Holzhaus and would feel more comfortable by not burning bridges behind us without any compelling reason.

Reply:
There is no compelling reason to maintain support for the legacy skin system with Qt6. Getting it to a releasable state would require massive refactoring and overcoming obstacles we never figured out in #1974.

Reply:
There's still a loooong way until QML skins can replace the legacy skins. Even if one QML skin is usable and polished that is no reason to drop the legacy skins.

Reply:
I strongly disagree. Keeping the legacy skin system means staying with Qt5.

Reply:
I'm not going to keep working on Qt6 support if you insist on making the legacy skin system build with Qt6 even though it will never fully work with Qt6.

Reply:
We will have CI building the legacy skin system with Qt5 until we're fully ready to switch to QML. Building the legacy skin system with Qt6 is pointless.

Reply:
I think there may be some misunderstanding here. This branch does not delete the legacy skin system. It just modifies CMakeLists.txt so the legacy skin system is only build with Qt5 but not Qt6.

Reply:
>  we're not going to use QOpenGLWidget.

Please elaborate why. I remember we were considering merging that PR 1 year ago. It may have some major issues and need refactoring, but that whole PR is <700 LoC. Even if getting it working properly needs 1K lines more, this is still a tiny fraction of the amount of code needed for QML.

I'm not saying that we have to get QOpenGLWidget working. We possibly don't. But you didn't yet elaborate why we have to decide this now.

I don't think it's wise to commit to remove the whole legacy skin system when switching to Qt6, without the possibility for any kind of transitional release, e.g. a release that supports Qt6 but stays with legacy skin system, or a release that supports both QML and legacy skin system side-by-side to iron out bugs.
It's too early to say if we need such a transitional release, but it would be nice to have this option. Removing everything for Qt6 just reduces our flexibility.

Reply:
Switching to QOpenGLWidget would be a huge waste of time and effort. We have no idea how to make it work. @rryan already tried quite hard to get it working well to no avail.

Reply:
The PR is old, Qt APIs may have changed to make it work.

The question is: Why is it not feasible to just disable the QGLWidget and disable the corresponding XML nodes in the legacy skin parser? Wouldn't that suffice to make Mixxx build with Qt6?

Reply:
The waveform code is much more complicated than that. Getting the legacy skin system building would require lots of `#if QT_VERSION` checks throughout the legacy code. I'm not wasting my time getting legacy code with no future building with Qt6.

Reply:
Did you actually verify it or is this statement pure speculation? It makes a difference.

Reply:
I have spent enough time with the waveform code to know how complex it is. I'm not touching it.

Reply:
Whether the legacy skin system could be built without waveforms and spinnies is a moot point. It won't be released in that state.

Reply:
> Whether the legacy skin system could be built without waveforms and spinnies is a moot point. It won't be released in that state.

It won't be released with an incomplete QML UI either. The point is that it's not completely unrealistic to get QOpenGLWidget working, and we should keep that option open.

Considering that the team just decided that we only want to have 3 month beta periods, the release that switches from Qt5 to Qt6 and replaces the whole UI of Mixxx in the same release, i'm pretty certain that release will bug-ridden and seriously hurt Mixxx' reputation as being very stable.

Reply:
It would help to have stripped but otherwise usable, temporary version for testing until the QML skin has a library integration.

Reversely we could re-enable legacy skin components gradually as needed by starting from a "buildable" state that pollutes the code only with the bare minimum of #ifdefs. Moving the affected files back in CMakeLists.txt is no big deal.

Reply:
If the other devs think this is a good idea, go ahead. But I think this is a mistake.

Reply:
> Moving the affected files back in CMakeLists.txt is no big deal.

Right, I don't understand why there is such a fuss about this. Moving a list of files around in CMakeLists.txt is trivial. Maintaining `#if QT_VERSION` checks all over the code is not.

Reply:
> it's not completely unrealistic to get QOpenGLWidget working

Yes it is. We tried and failed already. Even if we get it working, that would be seriously limiting Mixxx's performance by coupling it to OpenGL -- which Apple has declared deprecated! -- and continuing to render everything else on the CPU.

Reply:
> Considering that the team just decided that we only want to have 3 month beta periods, the release that switches from Qt5 to Qt6 and replaces the whole UI of Mixxx in the same release, i'm pretty certain that release will bug-ridden and seriously hurt Mixxx' reputation as being very stable.

We have agreed that betas will take as long as they need to. It would IMO be reasonable to expect the beta Qt5 -> Qt6 transition to be longer than the usual 3 months.

Reply:
> It would help to have stripped but otherwise usable, temporary version for testing until the QML skin has a library integration.

QML is already usable for testing by dragging and dropping tracks from a file browser. Inconvenient, but sufficient for testing.

Reply:
If we get scrolling waveforms in QML working, we could also replace QGLWidget with QQuickWidget containing just the scrolling waveforms. Its an ugly hack, but it might work.

Reply:
I would like to find a compromise that allows us to explore multiple directions concurrently. We should rule out an option only iff we hit issues that are *unresolvable* with any reasonable amount of effort.

@Holzhaus / @Swiftb0y Do you consider it feasible to start with this radical switch and then gradually try to re-enable the legacy skin components for Qt6? This would allow us to work in parallel. Does this PR introduce any irreversible changes that would prevent us from doing so?

@Be-ing We would need your consent that you support such a strategy and not undermine it by introducing new breaking changes without prior discussion.

Finding a compromise requires understanding of each others needs while patiently conveying your own ideas and aspirations. This is challenging and sometimes exhausting. Ignorance and denial is the worst strategy if you aim to convince someone.

“If you want to go fast, go alone. If you want to go far, go together.”

Reply:
> Do you consider it feasible to start with this radical switch and then gradually try to re-enable the legacy skin components for Qt6?

Yes, I advocated for this approach with the library but @Holzhaus was against it. However, I will not put any work into backtracking by putting QQuickWidgets into legacy skins.

Reply:
We cannot drop legacy code until the new code works fine. That's may whole point. Otherwise we'll have a broken mixxx main branch and cannot do a mixxx release in years.

I know this is just for Qt6 support and I'm okay with merging initial Qt6 support without the legacy skin system, but we should not burn bridges unless we need to.

> I want to implement a new feature on top of #2618 for one big effect chain with more effects, but just the thought of adding more complexity with the legacy skin system -- and repeating that four times! -- is making me reconsider whether that is worthwhile at this time.

I have no objections against a new hidden feature that is not accessible from the legacy skins, just from QML. But if that feature breaks legacy skins, we cannot merge at this time. QML skin feature parity is probably still years away, and I'd like to do more releases until then.

Reply:
We need to ensure that we are all in a position to make progress. No one is forced to work on features that they do not support. Nevertheless we need to coordinate our work and be aware of the differing goals to prevent conflicts and frustration.

I guess we all agree that we want and need to replace the legacy skins with QML asap. If it turns out that keeping the legacy skins alive imposes too much work or would pollute the code base unacceptably we can decide to stop this maintenance effort at any time and shift focus. If we could agree to proceed with the intent to defer this decision until it becomes inevitable instead of giving up recklessly this would help.

How about extracting Qt6-only components/modules into separate files if needed, e.g. the new QML UI based on #2618? This will hopefully prevent to mix conflicting Qt5/Qt6 code in the same file when managing the differences becomes inconvenient. Not sure how to deal with the split at the QML-level where we also have backwards compatible components that are not even using all the features available in Qt 5.15, mainly due to outdated Ubuntu dependencies.

Reply:
> We cannot drop legacy code until the new code works fine.

I fully agree. But wasting effort on maintaining legacy code with Qt6 does nothing to help us get there, it only puts obstacles in the way.

> I know this is just for Qt6 support and I'm okay with merging initial Qt6 support without the legacy skin system, but we should not burn bridges unless we need to.

The point of this approach is that it does not touch the legacy code at all. It just doesn't build the old code with Qt6. No bridges were harmed in the making of this PR.

Reply:
> Not sure how to deal with the split at the QML-level where we also have backwards compatible components that are not even using all the features available in Qt 5.15, mainly due to outdated Ubuntu dependencies.

When we get CI setup for Qt6 we could make QML build for only Qt6 and not worry about old Qt5 versions. Here is a PR to start that:
https://github.com/mixxxdj/vcpkg/pull/19

After we get vcpkg working on macOS, it should be easy to use vcpkg for Linux CI as well.